test/integration/server.h

```diff
@@ -151,6 +152,59 @@ class TestScopeWrapper : public Scope {
   ScopePtr wrapped_scope_;
 };
 
+// A counter which signals on a condition variable when it is incremented.
+class NotifyingCounter : public Stats::CounterImpl {
+public:
+  NotifyingCounter(Stats::StatName name, Stats::AllocatorImpl& alloc,
+                   Stats::StatName tag_extracted_name,
+                   const Stats::StatNameTagVector& stat_name_tags, absl::Mutex& mutex)
+      : Stats::CounterImpl(name, alloc, tag_extracted_name, stat_name_tags), mutex_(mutex) {}
+
+  void add(uint64_t amount) override {
+    // We don't actually need to log for the add, but we do need to make sure
+    // the mutex.Await() will trigger.
+    absl::MutexLock l(&mutex_);
+    Stats::CounterImpl::add(amount);
+  }
+
+private:
+  absl::Mutex& mutex_;
+};
+
+// A stats allocator which creates NotifyingCounters rather than regular CounterImpls.
+class NotifyingAllocatorImpl : public Stats::AllocatorImpl {
+public:
+  using Stats::AllocatorImpl::AllocatorImpl;
+
+  virtual Stats::CounterImpl* makeCounterImpl(StatName name, StatName tag_extracted_name,
+                                              const StatNameTagVector& stat_name_tags) {
+    Stats::CounterImpl* counter =
+        new NotifyingCounter(name, *this, tag_extracted_name, stat_name_tags, mutex_);
+    {
+      absl::MutexLock l(&mutex_);
+      counters_.emplace(counter->name(), counter);
+    }
+    return counter;
+  }
+
+  // Allow getting the counter directly from the allocator, since it's harder to
+  // signal when the counter has been added to a given stats store.
+  virtual Stats::CounterImpl* getCounter(std::string name) {
```

Comment:
You could also make this call block until the counter is created via an await. There's a bunch of similar code in TestStore in test/common/stats/stat_test_utility.h -- it'd be nice to re-use that somehow.

Also I'm wondering how you'd feel about wrapping and delegating rather than changing the underlying allocator...it's a bit more work because you'd have to making forwarding methods but also less intrusive to prod code.

From a practical perspective I have a preference for not putting significant prod impls in headers as it makes debugging a little messier as the code gets replicated in multiple object files, but it's not a deal breaker. There are other ways around that though.

At a minimum call this getCounterFromString per convention in that file.

Reply:
I realize I've hinted at "other ways around this" and should just say what I mean:

allocator_impl.h is somewhat broadly included, transitively. I think we can have allocator_impl_private.h which has these inline template methods in them, and include those only in allocator_impl.cc in prod, plus in a new cc file implementing your notiying allocator.

But LMK what you know about delegating in TestStore (test/common/stats/stat_test_utility.h), rather than subclassing the allocator for this. Maybe that won't be the right integration point for integration tests?

Reply:
I don't think that would work.
At the end of the day we need to know when the counter has been incremented.  AFIK that can only be done by either refactoring the actual production code to increment through the store (rather than getting the counter and incrementing directly) or subclassing the counter.

Reply:
Will look into wrapping and keeping the base classes in the .cc file.  If I do that work would you be OK with condvar for now, and you could clean it up once your mutex await fix lands?